Interior bcs

wence- opened the pull request:
Remove support for `method="geometric"` boundary conditions and then augment the bc construction such that we can handle bcs on interior marked facets. 

@ScottMacLachlan does this work for your use cases? I have only proved it correct, not tested.

wence- commented on firedrake/functionspacedata.py:
```diff
@@ -291,24 +269,23 @@ def get_top_bottom_boundary_nodes(mesh, key, V):
-    :arg key: The key a 3-tuple of ``(entity_dofs_key, sub_domain, method)``.
-        Where sub_domain indicates top or bottom and method is whether
-        we should identify dofs on facets topologically or geometrically.
+    :arg key: The key a 3-tuple of ``(entity_dofs_key, sub_domain)``.
```
2-tuple

sv2518 commented:
I think we should change the docs on "Boundary conditions in discontinuous spaces", where we advise to use geometric boundary conditions, within this PR as well. It's line 568 in https://github.com/firedrakeproject/firedrake/blob/wence/feature/no-geom-bcs/docs/source/variational-problems.rst.

ReubenHill commented:
I see I've been asked to review - I don't actually know much about how boundary conditions work in firedrake so don't quite feel I am able to comment usefully at the moment. Perhaps if you were to make Sophia's suggested change I might be able to start somewhere with the review?

It's also now quite clear to me _why_ this change as been made in either the description of the PR or the commit that makes the depreciation. Could you please add some context?

wence- commented:
@ScottMacLachlan have you had a chance to look at this? It would even be useful to have a testcase that you've been using.

ScottMacLachlan commented:
I now get an error in my call to MeshHierarchy, on a mesh read in from gmesh:

```Traceback (most recent call last):
  File "simpleTestSandwich.py", line 52, in <module>
    mh = MeshHierarchy(base, 1)
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/src/firedrake/firedrake/mg/mesh.py", line 140, in MeshHierarchy
    meshes = [mesh] + [mesh_builder(dm, dim=mesh.ufl_cell().geometric_dimension(),
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/src/firedrake/firedrake/mg/mesh.py", line 140, in <listcomp>
    meshes = [mesh] + [mesh_builder(dm, dim=mesh.ufl_cell().geometric_dimension(),
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/lib/python3.8/site-packages/decorator.py", line 231, in fun
    return caller(func, *(extras + args), **kw)
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/src/PyOP2/pyop2/profiling.py", line 60, in wrapper
    return f(*args, **kwargs)
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/src/firedrake/firedrake/mesh.py", line 1711, in Mesh
    topology = MeshTopology(plex, name=name, reorder=reorder,
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/lib/python3.8/site-packages/decorator.py", line 231, in fun
    return caller(func, *(extras + args), **kw)
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/src/PyOP2/pyop2/profiling.py", line 60, in wrapper
    return f(*args, **kwargs)
  File "/gpfs/fs0/scratch/s/scottmac/scottmac/April_2021/firedrake/src/firedrake/firedrake/mesh.py", line 700, in __init__
    dmcommon.label_facets(plex, label_boundary=label_boundary)
  File "firedrake/cython/dmcommon.pyx", line 994, in firedrake.cython.dmcommon.label_facets
RuntimeError: 63
```

Any ideas?

wence- commented:
Can you provide the .geo?

ScottMacLachlan commented:
> Can you provide the .geo?

Here it is (extension changed to allow attachment)...
[sandwichUniform_h=005_geo.txt](https://github.com/firedrakeproject/firedrake/files/6325460/sandwichUniform_h.005_geo.txt)

wence- commented:
This works forr me. I think that you didn't rebuild the Cython extensions in Firedrake after checking out the branch? If you have a separate PETSc, run `make` in the firedrake source directory, otherwise run `firedrake-update` after selecting this branch.

ScottMacLachlan commented:
You're right (of course).  After running make, I get identical output for my test problem using `DirichletBC` to what I got using the old `InteriorBC` workaround.  Thanks!

ScottMacLachlan commented:
FWIW, most of the code changes are in places where I'm not qualified to judge correctness.  What I do understand, however, looks good to me...

wence- commented:
> I think we should change the docs on "Boundary conditions in discontinuous spaces", where we advise to use geometric boundary conditions, within this PR as well. It's line 568 in https://github.com/firedrakeproject/firedrake/blob/wence/feature/no-geom-bcs/docs/source/variational-problems.rst.

I did this, thanks.

sv2518 reviewed (dismissed):
LGTM

wence- commented:
Added a very simple test @ScottMacLachlan do you want to sanity-check?

I also updated the triplot function to colour interior mesh markers as well, @danshapero does that look sane?

danshapero commented:
The plotting changes look good. I also tried it out with an unstructured quad mesh and everything looked fine there too.
